streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/TaskMovement.java

```diff
@@ -16,128 +16,94 @@
-            final UUID source = sourceClientEntry.getKey();
+        int remainingWarmupReplicas = maxWarmupReplicas;
+        for (final TaskMovement movement : taskMovements) {
+            final UUID leastLoadedClient = clientsByTaskLoad.poll(movement.task);
+            if (leastLoadedClient == null) {
+                throw new IllegalStateException("Tried to move task to caught-up client but none exist");
+            }
 
-            final Iterator<TaskId> sourceClientTasksIterator = sourceClientEntry.getValue().iterator();
-            while (sourceClientTasksIterator.hasNext()) {
-                final TaskId task = sourceClientTasksIterator.next();
-                final UUID destination = taskToDestinationClient.get(task);
-                if (destination == null) {
-                    log.error("Task {} is assigned to client {} in initial assignment but has no owner in the final " +
-                                  "balanced assignment.", task, source);
-                    throw new IllegalStateException("Found task in initial assignment that was not assigned in the final.");
-                } else if (!source.equals(destination)) {
-                    if (destinationClientIsCaughtUp(task, destination, tasksToCaughtUpClients)) {
-                        sourceClientTasksIterator.remove();
-                        statefulActiveTaskAssignment.get(destination).add(task);
-                    } else {
-                        if (clientStates.get(destination).prevStandbyTasks().contains(task)
-                                && tasksToRemainingStandbys.get(task) > 0
-                        ) {
-                            decrementRemainingStandbys(task, tasksToRemainingStandbys);
-                        } else {
-                            --remainingAllowedWarmupReplicas;
-                        }
+            final ClientState sourceClientState = clientStates.get(leastLoadedClient);
+            sourceClientState.assignActive(movement.task);
 
-                        movements.add(new TaskMovement(task, source, destination));
-                        if (remainingAllowedWarmupReplicas == 0) {
-                            return movements;
-                        }
-                    }
-                }
+            final ClientState destinationClientState = clientStates.get(movement.destination);
+            if (destinationClientState.prevStandbyTasks().contains(movement.task) && tasksToRemainingStandbys.get(movement.task) > 0) {
+                decrementRemainingStandbys(movement.task, tasksToRemainingStandbys);
+                destinationClientState.assignStandby(movement.task);
+                warmupReplicasAssigned = true;
+            } else if (remainingWarmupReplicas > 0) {
+                --remainingWarmupReplicas;
+                destinationClientState.assignStandby(movement.task);
+                warmupReplicasAssigned = true;
             }
```

Comment:
Hm. I'm not sure this was what you meant, but I think we have to interpret "`wasn't previously assigned to any client"` as meaning `"any client other than the source client"`. By definition someone had to have had this task previously for it to be involved in a movement: of course this could be due to leftover/old state and not the actual assigned task, but we currently don't and can't distinguish these.

Given that, I think I buy this. But should the condition also be generalized to "the task had fewer than num.standby previous clients" (after accounting for the above)?

Just to lay out the general reasoning for my  future self: we basically only want to count something as a warmup replica when counting it against the total standbys for this rebalance(s) would mean temporarily revoking a standby task from some client for the duration of the rebalance(s). In other words, outside of the tasks & clients involved in a movement, we want the assignment during the intermediate rebalances to resemble the final assignment as much as possible.

So if we had num.standbys = 5 and 4 previous clients I think we would still want to consider this a standby. Of course, nothing discussed here will really matter unless (until) we make the standby assignment more sticky and/or lag-based

Reply:
Hmm, it doesn't _seem_ like there's such a guarantee. But I'm always suspicious I just mised the point.

What if we had lost the node holding the standby? Or we lost the node holding the active, and decided to promote the standby to active? Or if we were running with no standbys and the setting were adjusted to allow standbys? Or something similar... It seems like in these situations, there is no prior owner of the standby, and (because the cluster changed) we might want to move the active to a new node, and there wouldn't be a valid candidate to host the standby, so we might as well let the destination _be_ the standby without using up a "warmup" task.

Reply:
I suspect this may also be related to the issue we just resolved, ie we know all tasks here must have caught-up clients. But just to clarify what I meant in the first paragraph, this client may be caught-up due to old leftover state rather than a previous task. Unfortunately(?) we just encode the offset sums for everything, so we can't distinguish these cases. 

I think we're actually in agreement here, but I'd propose a slight relaxation of the constraint: if N is the number of clients who previously had this task as a standby, let we count it as a standby if N < num.standby.replicas. This includes the cases you described, but also extends to ones where we have multiple standbys and lost a single node. WDYT?

Reply:
Yeah, something like that sounds right to me. I think what we have here is "good enough" for now, though, and we can merge this PR while continuing to refine the standby accounting later.